tiledb/sm/subarray/subarray.cc

```diff
@@ -1016,6 +1024,103 @@ Status Subarray::set_est_result_size(
+      break;
+    case Datatype::UINT8:
+      coalesced = coalesce_range<uint8_t>(dim_idx, range);
+      break;
+    case Datatype::INT16:
+      coalesced = coalesce_range<int16_t>(dim_idx, range);
+      break;
+    case Datatype::UINT16:
+      coalesced = coalesce_range<uint16_t>(dim_idx, range);
+      break;
+    case Datatype::INT32:
+      coalesced = coalesce_range<int32_t>(dim_idx, range);
+      break;
+    case Datatype::UINT32:
+      coalesced = coalesce_range<uint32_t>(dim_idx, range);
+      break;
+    case Datatype::INT64:
+      coalesced = coalesce_range<int64_t>(dim_idx, range);
+      break;
+    case Datatype::UINT64:
+      coalesced = coalesce_range<uint64_t>(dim_idx, range);
+      break;
+    case Datatype::FLOAT32:
+    case Datatype::FLOAT64:
+      // We can not reasonably coalesce floating point types.
+      break;
+    case Datatype::DATETIME_YEAR:
+    case Datatype::DATETIME_MONTH:
+    case Datatype::DATETIME_WEEK:
+    case Datatype::DATETIME_DAY:
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+      coalesced = coalesce_range<int64_t>(dim_idx, range);
+      break;
+    case Datatype::CHAR:
+      coalesced = coalesce_range<char>(dim_idx, range);
+      break;
+    case Datatype::STRING_ASCII:
+    case Datatype::STRING_UTF8:
+    case Datatype::STRING_UTF16:
+    case Datatype::STRING_UTF32:
+    case Datatype::STRING_UCS2:
+    case Datatype::STRING_UCS4:
+      // We can not reasonably coalesce string types.
+      break;
+    case Datatype::ANY:
+      coalesced = coalesce_range<uint8_t>(dim_idx, range);
+      break;
+    default:
+      assert(false);
+  }
+
+  // If we did not coalesce `range` into an existing range in
+  // `ranges_`, append it to `ranges_` as its own range.
+  if (!coalesced)
+    ranges_[dim_idx].emplace_back(range);
+}
+
+template <class T>
+bool Subarray::coalesce_range(const uint32_t dim_idx, const Range& range) {
+  // If range coalescing has been disabled, we're done.
+  if (!coalesce_ranges_)
+    return false;
+
+  std::vector<Range>* const ranges = &ranges_[dim_idx];
+
+  // If `ranges` is empty, there is not an existing range to coalesce with.
+  if (ranges->empty())
+    return false;
+
+  // If the start index of `range` immediately proceeds the end of the
+  // last range on `ranges`, they are contiguous and will be coalesced.
+  Range* const last_range = &ranges->back();
+  const bool contiguous = *static_cast<const T*>(last_range->end()) + 1 ==
+                          *static_cast<const T*>(range.start());
+
+  // Coalesce `range` with `last_range` if they are contiguous.
+  if (contiguous)
+    *static_cast<T*>(last_range->end()) = *static_cast<const T*>(range.end());
```

Comment:
This assumed fixed-sized ranges. Perhaps turn this into a proper setter (`set_end`) which updates also updates the range size? That would probably be a bit cleaner / safer?

Reply:
Done. We should only be coalescing fixed types. I've added a setter that will fatal if invoked on a var-sized range.